chunked_stream/lib/src/chunked_stream_iterator.dart

```diff
@@ -23,197 +23,179 @@ import 'dart:math';
-      _wakeup();
-    }, onError: (e, st) {
-      _error = e;
-      _stackTrace = st;
-      _wakeup();
-    });
-    _subscription.pause();
+abstract class ChunkedStreamIterator<T> {
+  factory ChunkedStreamIterator(Stream<List<T>> stream) {
+    return _ChunkedStreamIterator<T>(stream);
   }
 
-  /// Create a [ChunkedStreamIterator] over [stream].
-  factory ChunkedStreamIterator(Stream<List<T>> stream) =>
-      ChunkedStreamIterator._(stream);
+  /// Returns a list of the next [size] elements in original chunks. The size of
+  /// the chunks are not guaranteed to be consistent.
+  ///
+  /// Returns a list with a total of less than [size] elements split in chunks
+  /// if the end of stream is encounted before [size] elements are read.
+  ///
+  /// If an error is encountered before reading [size] elements, the error
+  /// will be thrown.
+  Future<List<List<T>>> readPreservingChunks(int size);
 
-  void _reset() {
-    _wakeup = null;
-    if (!_subscription.isPaused) {
-      _subscription.pause();
-    }
-  }
+  /// Returns a list of the next [size] elements.
+  ///
+  /// Returns a list with less than [size] elements if the end of stream is
+  /// encounted before [size] elements are read.
+  ///
+  /// If an error is encountered before reading [size] elements, the error
+  /// will be thrown.
+  Future<List<T>> read(int size);
 
-  /// Read a series of chunks from the stream until we have [size] number of
-  /// items, then return these items.
+  /// Cancels the stream iterator (and the underlying stream subscription)
+  /// early.
   ///
-  /// This returns less than [size], if end of stream occurs before [size]
-  /// number of items have been received.
+  /// The [ChunkedStreamIterator] is automatically cancelled if [read] or
+  /// [readPreservingChunks] reaches the end of the stream or an error.
   ///
-  /// If an error occurs before receiving [size] items, the error will be thrown
-  /// and next call to [read] will with data already buffered from before the
-  /// error occurred.
+  /// Users should call [cancel] to ensure that the stream is properly closed
+  /// if they need to stop listening earlier than the end of the stream.
+  Future<void> cancel();
+
+  /// Creates a sub-[Stream] with the next [size] elements.
   ///
-  /// This method may not be called concurrently.
-  Future<List<T>> read(int size) async {
-    ArgumentError.checkNotNull(size, 'size');
-    if (size <= 0) {
-      throw ArgumentError.value(size, 'size', 'size must be greater than zero');
-    }
-    if (_wakeup != null) {
-      throw StateError('Concurrent invocations not supported');
-    }
+  /// The resulting stream may contain less than [size] elements if the
+  /// underlying stream has less than [size] elements before the end of stream.
+  ///
+  /// If [read] or [readPreservingChunks] is called before the sub-[Stream] is
+  /// fully read, the remainder of the elements in the sub-[Stream] will be
+  /// automatically drained.
```

Comment:
Would it not be preferable to make this an error?

We already state that the API cannot be called concurrently. I think it would be fair to elaborate that calling `read` or `substream` before a previously returned substream has been drained, consumed or cancelled throws.

IMO, we should throw a `StateError` in this case.

-----------------

I think we should consider making this an error rather than a silent draining, imagine: I'm reading a tar-file... I pass the substream off to something that sends over the network... But for some reason I forget to await that substream is fully sent before I read the next entry from the tar-reader.. Depending on how events a propagated and how fast my loop gets back to reading the next entry, it's possible that the thing sending my substream will end up sending a partial substream, right? (because next time I try to read an entry, my previous substream is drained).

I suspect you might want to say that, if nobody has started listening on the substream before the next call to `read`, then it is drained. But even in this case, I suspect it's preferable and more predictable to throw a `StateError` and tell the user to `.drain()` or cancel their subscription, before they call `read` again.